File: src/data/sort.ts
export interface CostTokenSortable {
  tokens: number;
  costUSD?: number;
  cost?: number | null;
}

export function costValue(entry: CostTokenSortable): number {
  return entry.costUSD ?? entry.cost ?? 0;
}

export function compareByCostThenTokens(a: CostTokenSortable, b: CostTokenSortable): number {
  return costValue(b) - costValue(a) || b.tokens - a.tokens;
}

File: crates/aitrack-core/src/data/sort.rs
//! Sorting utilities for usage data.

use super::aggregate::ModelAgg;
use std::cmp::Ordering;

/// Compare two ModelAgg entries by cost (descending).
pub fn compare_by_cost(a: &ModelAgg, b: &ModelAgg) -> Ordering {
    b.cost_usd
        .partial_cmp(&a.cost_usd)
        .unwrap_or(Ordering::Equal)
}

/// Compare two ModelAgg entries by input tokens (descending).
pub fn compare_by_input(a: &ModelAgg, b: &ModelAgg) -> Ordering {
    b.input_tokens.cmp(&a.input_tokens)
}

/// Compare two ModelAgg entries by output tokens (descending).
pub fn compare_by_output(a: &ModelAgg, b: &ModelAgg) -> Ordering {
    b.output_tokens.cmp(&a.output_tokens)
}

/// Compare two ModelAgg entries by number of days (descending).
pub fn compare_by_days(a: &ModelAgg, b: &ModelAgg) -> Ordering {
    b.days.cmp(&a.days)
}
